src/llm/visual_language_model/servable.cpp

```diff
@@ -41,24 +41,37 @@ std::shared_ptr<GenAiServableProperties> VisualLanguageModelServable::getPropert
     return properties;
 }
 
+bool VisualLanguageModelServable::supportsSpeculativeDecoding() const {
+    return false;
+}
+
 absl::Status VisualLanguageModelServable::prepareInputs(std::shared_ptr<GenAiServableExecutionContext>& executionContext) {
     auto vlmExecutionContext = std::static_pointer_cast<VisualLanguageModelServableExecutionContext>(executionContext);
-    // Earlier we should validate that chat completion endpoint is used.
     if (vlmExecutionContext->apiHandler == nullptr) {
         return absl::Status(absl::StatusCode::kInvalidArgument, "API handler is not initialized");
     }
-
-    vlmExecutionContext->inputImages = vlmExecutionContext->apiHandler->getImages();
-
-    auto& chatHistory = vlmExecutionContext->apiHandler->getChatHistory();
-    if (chatHistory.size() == 0) {
-        return absl::Status(absl::StatusCode::kInvalidArgument, "Chat history is empty");
+    if (executionContext->endpoint == Endpoint::CHAT_COMPLETIONS) {
+        ov::genai::ChatHistory& chatHistory = vlmExecutionContext->apiHandler->getChatHistory();
+        ImageHistory& imageHistory = vlmExecutionContext->apiHandler->getImageHistory();
+        size_t imageIndex = 0;
+        for (auto& image : imageHistory) {
+            auto& [chatTurnIndex, imageTensor] = image;
+            std::string imageTag = "<ov_genai_image_" + std::to_string(imageIndex++) + ">\n";
+            if (chatHistory[chatTurnIndex].find("content") != chatHistory[chatTurnIndex].end()) {
```

Comment:
undefined behavior when chatHistory does not contain chatTurnIndex

Reply:
It's not external input. `chatTurnIndex` is prepared along with `chatHistory` in API handler. 
We can check it, but it feels like we don't trust our own implementation.